[qos]New headroom calculation

stephenxs opened the pull request:
We're going to improve the way we calculating headroom size. Originally we generate headroom size by looking up the predefined lookup table.
We will have the following groups of parameters
- List of SONiC configuration (such as speed and cable length)
- List of pg configuration (such as cell size, IPG, MTU, Peer response time, small packet size percentage)

Based on the parameters and a well-known formula the code in buffer manager will do the calculation and not take it from pre-defined values as we have today.
On top of that, need an ability to override headroom and do not calculate it in the code.

Commit:
[qos]New headroom calculation -- initial version

keboliu commented on doc/qos/new-headroom-calculation.md:
```diff
@@ -0,0 +1,168 @@
+- When system start, it reads the pg_profile_lookup.ini and generate an internal lookup table indexed by speed and cable length, and containing size, xon, xoff and threshold.
+- When a port's cable length updated, it records the cable length of the port. But it doesn't update relavent tables accordingly.
+- When a port's speed updated,
+    1. It looks the (speed, cable length) pair up in the BUFFER_PROFILE table or generate a new profile according to the internal lookup table.
```
looks --> "looks up"

keboliu commented on doc/qos/new-headroom-calculation.md:
```diff
@@ -0,0 +1,168 @@
+## Current Solution
+
+- When system start, it reads the pg_profile_lookup.ini and generate an internal lookup table indexed by speed and cable length, and containing size, xon, xoff and threshold.
+- When a port's cable length updated, it records the cable length of the port. But it doesn't update relavent tables accordingly.
```
are you saying there is a bug here? "it doesn't update the relevant tables?"

stephenxs replied:
IMO, it's a bug. The BUFFER_PG table should be updated accordingly to the port's cable length and speed after they have been updated.

keboliu commented on doc/qos/new-headroom-calculation.md:
```diff
@@ -0,0 +1,168 @@
+- List of pg configuration (such as cell size, IPG, MTU, Peer response time, small packet size percentage)
+
+Based on the parameters and a well-known formula the code in buffer manager will do the calculation and not take it from a pre-defined values as we have today.
+On top of that, we need to support the ability to override headroom and not to calculate it in the code.
```
can you elaborate "override headroom"?

stephenxs replied:
To enable user to configure a static headroom size instead of dynamically calculate it.

keboliu commented on doc/qos/new-headroom-calculation.md:
```diff
@@ -0,0 +1,168 @@
+On top of that, we need to support the ability to override headroom and not to calculate it in the code.
+# Design
+## The overall flow
+
```
can we have a flow chart here?

stephenxs replied:
will add.

keboliu commented on doc/qos/new-headroom-calculation.md:
```diff
@@ -0,0 +1,168 @@
+# Design
+## The overall flow
+
+ - When system starts, load the SWITCH_PARAMETER table or the pg_profile_lookup.ini.
```
I thought we are removing "pg_profile_lookup.ini"?

stephenxs replied:
But other vendors probably aren't able to provide SWITCH_PARAMETERs in one day.
Let's keep it in order to tolerance vendors using the old approach to calculating the headroom.

keboliu commented on doc/qos/new-headroom-calculation.md:
````diff
@@ -0,0 +1,168 @@
+    type        = "static" / "dynamic"      ; Optional. Whether the profile is dynamically calculated or user configured.
+                                            ; Default value is "static"
+```
+The `static` profile is configured by CLI while the `dynamic` profile is synamically calculated when system starts or a port's cable length or speed is updated.
````
synamically,  typo? I guess should be dynamically.

keboliu commented on doc/qos/new-headroom-calculation.md:
````diff
@@ -0,0 +1,168 @@
+```
+    type        = "static" / "dynamic"      ; Mandatory. dynamic means a dynamically calculated buffer profile is adopted
+                                            ; while static means a static profile is adopted for headroom override. 
+```
````
since there is DB schema change, do you need to consider DB migration?

stephenxs replied:
If we're going to downgrade from a version that supports those newly introduced fields to a version that doesn't, we need to.
Otherwise, we don't because the entries without those newly introduced fields will be treated as "dynamic" one.

keboliu commented on doc/qos/new-headroom-calculation.md:
````diff
@@ -0,0 +1,168 @@
+## Command line interface
+### To configure a static profile
+```
+sonic#config buffer_profile <name> add <xon> <xoff> <headroom> <dnyamic_th>
````
I suggest add more detail about the CLI here, Are all the parameters needed when creating a new profile or we can update part of them afterward?

stephenxs replied:
will add.

Commit:
[headroom calculation]updated with dynamic tables moved to APPL_DB

Commit:
[qos]Use CONFIG_DB instead of APPL_DB
Eventually, I decided to store buffer related tables in CONFIG_DB
because to store them in APPL_DB doesn't yeild to much benefits.
The discussion regarding CONFIG_DB vs APPL_DB is moved to the end of the
doc.

Commit:
[qos]Add sonic-to-sonic update flow and warm reboot flow

Commit:
[qos]Address review comments

Commit:
1. rewording.
2. add behavior of current solution

Commit:
[qos]Support configure lossless PG and corresponding handling.
Supply some commands.

Commit:
[qos]Address comments:
1. the dynamic th should remain for dynamically calculated entries
2. should check whether the headroom size is legal after speed and cable
length updated
3. extend command line interface
4. add test case
5. Use APPL_DB

Commit:
[qos]Update doc

Commit:
[qos]Adjust some flow.

stephenxs committed:
Address comments

Commit:
[qos] Optimize the flow.

Commit:
[qos] remove headroom_type in BUFFER_PG, move some maximum values into
another table and add SAI interface to fetch it.

Commit:
[qos] dynamic headroom calculation update
Add logic of handling buffer related CONFIG_DB
Add definition of lua plugin for calculating headroom and shared buffer pool and for checking legality of the accumulative headroom of a port
Fix typo in well-known formula
Correct the errors in command section.

Commit:
correct the key for BUFFER_MAX_PARAMETER
Fix some minor errors

Signed-off-by: Stephen Sun <stephens@mellanox.com>

Commit:
Update CLI

Signed-off-by: Stephen Sun <stephens@mellanox.com>
